File: app/src/main/java/de/westnordost/streetcomplete/data/osm/persist/RelationDao.java
package de.westnordost.streetcomplete.data.osm.persist;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import de.westnordost.streetcomplete.util.Serializer;
import de.westnordost.osmapi.map.data.OsmRelation;
import de.westnordost.osmapi.map.data.Relation;
import de.westnordost.osmapi.map.data.RelationMember;

public class RelationDao extends AOsmElementDao<Relation>
{
	private final Serializer serializer;
	private final SQLiteStatement insert;

	@Inject public RelationDao(SQLiteOpenHelper dbHelper, Serializer serializer)
	{
		super(dbHelper);
		this.serializer = serializer;

		String sql = "INSERT OR REPLACE INTO " + RelationTable.NAME + " ("+
				RelationTable.Columns.ID+","+
				RelationTable.Columns.VERSION+","+
				RelationTable.Columns.MEMBERS+","+
				RelationTable.Columns.TAGS+
				") values (?,?,?,?);";
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		insert = db.compileStatement(sql);
	}

	@Override protected String getTableName()
	{
		return RelationTable.NAME;
	}

	@Override protected String getIdColumnName()
	{
		return RelationTable.Columns.ID;
	}

	@Override protected String getElementTypeName()
	{
		return Relation.Type.RELATION.name();
	}

	@Override protected void executeInsert(Relation relation)
	{
		insert.bindLong(1, relation.getId());
		insert.bindLong(2, relation.getVersion());
		insert.bindBlob(3, serializer.toBytes(new ArrayList<>(relation.getMembers())));
		if(relation.getTags() != null)
		{
			HashMap<String, String> map = new HashMap<>(relation.getTags());
			insert.bindBlob(4, serializer.toBytes(map));
		}
		else
		{
			insert.bindNull(4);
		}

		insert.executeInsert();
		insert.clearBindings();
	}

	@Override protected Relation createObjectFrom(Cursor cursor)
	{
		int colId = cursor.getColumnIndexOrThrow(RelationTable.Columns.ID),
			colMembers = cursor.getColumnIndexOrThrow(RelationTable.Columns.MEMBERS),
			colVersion = cursor.getColumnIndexOrThrow(RelationTable.Columns.VERSION),
			colTags = cursor.getColumnIndexOrThrow(RelationTable.Columns.TAGS);

		long id = cursor.getLong(colId);
		int version = cursor.getInt(colVersion);
		Map<String,String> tags = null;
		if(!cursor.isNull(colTags))
		{
			tags = serializer.toObject(cursor.getBlob(colTags), HashMap.class);
		}
		List<RelationMember> members = serializer.toObject(cursor.getBlob(colMembers), ArrayList.class);

		return new OsmRelation(id, version, members, tags);
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/persist/RelationDao.kt
package de.westnordost.streetcomplete.data.osm.persist

import android.database.Cursor
import android.database.sqlite.SQLiteOpenHelper
import androidx.core.content.contentValuesOf
import de.westnordost.osmapi.map.data.*
import de.westnordost.streetcomplete.data.ObjectRelationalMapping

import java.util.ArrayList
import java.util.HashMap

import javax.inject.Inject

import de.westnordost.streetcomplete.util.Serializer
import de.westnordost.streetcomplete.data.osm.persist.RelationTable.Columns.ID
import de.westnordost.streetcomplete.data.osm.persist.RelationTable.Columns.MEMBERS
import de.westnordost.streetcomplete.data.osm.persist.RelationTable.Columns.TAGS
import de.westnordost.streetcomplete.data.osm.persist.RelationTable.Columns.VERSION
import de.westnordost.streetcomplete.data.osm.persist.RelationTable.NAME
import de.westnordost.streetcomplete.ktx.*

class RelationDao @Inject constructor(dbHelper: SQLiteOpenHelper, override val mapping: RelationMapping)
    : AOsmElementDao<Relation>(dbHelper) {

    override val tableName = NAME
    override val idColumnName = ID
    override val elementTypeName = Element.Type.RELATION.name
}

class RelationMapping @Inject constructor(private val serializer: Serializer)
    : ObjectRelationalMapping<Relation> {

    override fun toContentValues(obj: Relation) = contentValuesOf(
        ID to obj.id,
        VERSION to obj.version,
        MEMBERS to serializer.toBytes(ArrayList(obj.members)),
        TAGS to obj.tags?.let { serializer.toBytes(HashMap(it)) }
    )

    override fun toObject(cursor: Cursor) = OsmRelation(
        cursor.getLong(ID),
        cursor.getInt(VERSION),
        serializer.toObject<ArrayList<OsmRelationMember>>(cursor.getBlob(MEMBERS)) as List<RelationMember>,
        cursor.getBlobOrNull(TAGS)?.let { serializer.toObject<HashMap<String, String>>(it) }
    )
}
